source.go

```diff
@@ -43,18 +56,215 @@ func (s *Source) Configure(ctx context.Context, cfg map[string]string) error {
+func (s *Source) Ack(context.Context, sdk.Position) error {
 	return nil
 }
 
-func (s *Source) Teardown(ctx context.Context) error {
+func (s *Source) Teardown(context.Context) error {
 	return nil
 }
+
+func (s *Source) populateIDs(ctx context.Context) error {
+	if len(s.fetchIDs) > 0 {
+		return nil
+	}
+	// the first read attempt (when the connector starts)
+	if !s.firstFetch {
+		sdk.Logger(ctx).Debug().
+			Dur("poll_interval", s.config.pollInterval).
+			Msg("sleeping before checking for changes")
+		time.Sleep(s.config.pollInterval)
+	}
+	s.firstFetch = false
+
+	sdk.Logger(ctx).Debug().Msg("populating IDs")
+	fetch := true
+	var cursor notion.Cursor
+	for fetch {
+		results, err := s.getPages(ctx, cursor)
+		if err != nil {
+			return fmt.Errorf("search failed: %w", err)
+		}
+		s.addToFetchIDs(ctx, results)
+
+		fetch = results.HasMore
+		cursor = results.NextCursor
+	}
+
+	sdk.Logger(ctx).Info().Msgf("fetched %v IDs", len(s.fetchIDs))
+	return nil
+}
+
+func (s *Source) addToFetchIDs(ctx context.Context, results *notion.SearchResponse) {
+	for _, result := range results.Results {
+		switch result.GetObject().String() {
+		case "page":
+			page := result.(*notion.Page)
+			sdk.Logger(ctx).Trace().
+				Str("page_id", page.ID.String()).
+				Time("last_edited_time", page.LastEditedTime).
+				Time("created_time", page.CreatedTime).
+				Msg("checking if page has changed")
+			if s.hasChanged(page) {
+				s.fetchIDs = append(s.fetchIDs, page.ID.String())
+			}
+		default:
+			sdk.Logger(ctx).Warn().
+				Str("object_type", result.GetObject().String()).
+				Msg("object type currently not supported")
+		}
+	}
+}
+
+func (s *Source) hasChanged(page *notion.Page) bool {
+	// see discussion in docs/cdc.md
+	lastTopMinute := time.Now().Truncate(time.Minute)
+	return page.LastEditedTime.After(s.lastEditedTime) &&
+		page.LastEditedTime.Before(lastTopMinute)
+}
+
+func (s *Source) getPages(ctx context.Context, cursor notion.Cursor) (*notion.SearchResponse, error) {
+	req := &notion.SearchRequest{
+		StartCursor: cursor,
+		Sort: &notion.SortObject{
+			Direction: notion.SortOrderASC,
+			Timestamp: notion.TimestampLastEdited,
+		},
+		Filter: map[string]string{
+			"property": "object",
+			"value":    "page",
+		},
+	}
+	return s.client.Search.Do(ctx, req)
+}
+
+func (s *Source) blockToRecord(parent notion.Block, children notion.Blocks) (sdk.Record, error) {
+	nb := NotionBlock{
+		BlockType: parent.GetType().String(),
+		Children:  children,
+	}
+	payload, err := json.Marshal(nb)
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("failed marshalling payload: %w", err)
+	}
+
+	return sdk.Record{
+		Position:  s.getKey(parent),
+		Metadata:  nil,
+		CreatedAt: time.Now(),
+		Key:       sdk.RawData(parent.GetID().String()),
+		Payload:   sdk.RawData(payload),
+	}, nil
+}
+
+func (s *Source) getKey(b notion.Block) sdk.Position {
+	return sdk.Position(b.GetLastEditedTime().Format(time.RFC3339))
```

Comment:
Thinking out loud - could using this time format ever be the source of an issue in the connector? It contains the timezone for example, I wonder how this affects us when the machine changes to DST, or if the record is processed on another machine. I'm not sure if this is a problem or not, but I'd rather use a UNIX timestamp if you ask me, just to be safe. As a side benefit - it would probably be faster to parse / encode.

Reply:
I think you're right about the Unix timestamp, I'll do that. Thanks for the suggestion!